### PR TITLE
Fix tab completion for parameters so that it shows common parameters as available

### LINKED ISSUE #1265: Tab completion confusion with common parameters
## Steps to reproduce

`function foo { param($DebugBuild) }`
1. `Receive-Job -wa<TAB>`
2. `foo -D<TAB>`
3. `foo -Debug<TAB>`
## Expected behavior
1. Cycle through -Wait and -WarningAction
2. Cycle through -Debug and -DebugBuild
3. Cycle through -Debug and -DebugBuild
## Actual behavior
1. Only completes -WarningAction
2. Only completes -DebugBuild
3. No additional completion

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

The current code deliberately filtered out common parameters and preferring specific parameters.  This resulted in tab completion where you only got results if the text matched a cmdlet parameter even though it would also match a common parameter.  Fix is to use the filtering code only if it's not a tab complete situation.  The `tryExactMatching` is only used when the cmdlet is being executed and not when tab complete is being used so we can use that to only exercise the filtering code.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/1265

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/MergedCommandParameterMetadata.cs [hunk 1]
```diff
@@ -500,47 +500,27 @@ internal string GetParameterSetName(uint parameterSet)
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/MergedCommandParameterMetadata.cs]: @SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test completion with exact match
```powershell
Expected a collection with size 3, but got collection with size 4 @(System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult).
at <ScriptBlock>, /Users/vsts/agent/2.158.0/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches | Should -HaveCount 3
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/MergedCommandParameterMetadata.cs]: @SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-windows`
WSMan Config Provider tab complete tests.Tab completion gets dynamic parameters for '<empty>' using '-co'
```powershell
Expected a collection with size 1, but got collection with size 2 @(System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult).
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1363
1363:         $res.CompletionMatches | Should -HaveCount $expected.Count
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/MergedCommandParameterMetadata.cs]: @SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test completion with exact match
```powershell
Expected a collection with size 3, but got collection with size 4 @(System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult, System.Management.Automation.CompletionResult).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches | Should -HaveCount 3
```

**Comment by iSazonov** [comment]
I guess that initial design was to hide common parameters because in cmdlets with many parameters they can annoy. If I remember right there was a suggestion to show common parameters with<Ctrl-Tab> but hide with <Tab>.

**Comment by SteveL-MSFT** [comment]
With the new update and simpler fix, this is not a breaking change :)

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 2]
```diff
@@ -895,6 +895,20 @@ dir -Recurse `
...
+            $res = TabExpansion2 -inputScript $inputStr -cursorColumn $inputStr.Length
+            $res.CompletionMatches | Should -HaveCount 3
+            [string]::Join(',', ($res.CompletionMatches.completiontext | Sort-Object)) | Should -BeExactly "-OutBuffer,-OutFile,-OutVariable"
+        }
+
+        It "Test completion with exact match" {
+            $inputStr = 'get-content -wa'
+            $res = TabExpansion2 -inputScript $inputStr -cursorColumn $inputStr.Length
+            $res.CompletionMatches | Should -HaveCount 4
+            [string]::Join(',', ($res.CompletionMatches.completiontext | Sort-Object)) | Should -BeExactly "-wa,-Wait,-WarningAction,-WarningVariable"
```
Comment (iSazonov) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: WarningVariable is a common variable too.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1 [hunk 3]
```diff
@@ -47,13 +47,17 @@ Describe "Get-Member" -Tags "CI" {
         $results[0].Name | Should -BeExactly 'Size'
         $results[1].Name | Should -BeExactly 'Zero'
     }
+
+    It "Should work with incomplete parameter '-i'" {
+        $a = 1
+        Get-Member -i $a | Should -Not -BeNullOrEmpty
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1]: Maybe:
```suggestion
        Get-Member -i $a | Should -Not -Throw
```

Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1]: Would need scriptblock braces as well but 👍

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 (hunk 2 above)
Comment (SteveL-MSFT) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: Yes, that is included in this validation.  Not sure what you are referring to.

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Windows`

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: Sorry, I meant previous test "Test completion with common parameters" - looks like duplicate.

Comment (SteveL-MSFT) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: I see.  This test is to ensure that even if provided an exact match `-wa` is an alias to `-warningaction` so it's an exact match, that you still get other completions.  Currently, if you do `get-content -wa`, you only get `-wa` and `-warningaction` and not the others.

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry static

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-static-analysis`

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**